### PR TITLE
Remove Tempfile

### LINKED ISSUE #5939: Remove class Tempfile
`Tempfile` is actually identical to `File` except for the constructor (which can be put in `File` with a different name), a few class methods (where this is also true), and the instance methods `delete` and `unlink` which are just aliases to `File.delete(@path)`.

I don't see a point in having a special type for this as the behaviour is identical to `File`, it's just an alternate way of initializing the instance.

The class methods/constructors could be renamed as followed:

* `Tempfile.new` -> `File.tempfile`
* `Tempfile.open` -> `File.tempfile` (yield)
* `Tempfile.dirname` -> `Dir.tempdir`
* `Tempfile.tempname` -> `File.tempname`

This issue is loosely related to #5938 about adjusting the arguments to the class methods/constructors but that's independent of the actual names.

### PR DESCRIPTION
**Pull request by straight-shoota**
This removes `Tempfile` and places its methods in `File` and `Dir` as proposed in #5939 and standardizes the arguments as proposed in #5938. All tempfile methods now receive `prefix`, `suffix` (and optional `dir`) arguments. There are also overload for only one argument `suffix` (and optional `dir`) because it is pretty common to create a temp file with a specific extension.

Previously, `require "tempfile"` was needed in order to use this feature. Now that it is integrated into `File` I think it makes sense to have it in corelib.

I'm not sure if `tempfile` with block should perhaps actually delete the tempfile. Or if there could be some other mechanism to simplify automatic deletion of tempfiles. This could be discussed in a follow-up issue, though.
Same goes for the addition of a method to create a temporary directory (see #4096). The only issue with this is that `Dir.tempdir` would also be a fitting name for this but it is now used to query the tempdir environment setting.
Related to that, there could also be a future improvement to allow modifying the global tempdir, probably in a way similar to `Time::Location.local`.

Fixes #5939 and #5938

### CODE REVIEW

File: src/file/tempfile.cr [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  # These values may contain directory separators.
+  #
+  # The file will be placed in *dir* which defaults to the standard temporary directory `Dir.tempdir`.
+  #
+  # *encoding* and *invalid* are passed to `IO#set_encoding`.
+  #
+  # It is the caller's responsibility to remove the file when no longer needed.
+  def self.tempfile(prefix : String?, suffix : String?, *, dir : String = Dir.tempdir, encoding = nil, invalid = nil)
+    fileno, path = Crystal::System::File.mktemp(prefix, suffix, dir)
+    new(path, fileno, blocking: true, encoding: encoding, invalid: invalid)
```
Comment (asterite) [inline src/file/tempfile.cr]: Why blocking true?

Comment (straight-shoota) [reply src/file/tempfile.cr]: Not sure... I just copied it form `Tempfile#initialize`.

**Comment by ysbaddaden** [comment]
As said in issues about this, I don't like this breaking change. I prefer to keep concerns distinct, and really see no need to push this into corelib.

**Comment by straight-shoota** [comment]
I think we established that there is little reason for `Tempfile` being a separate type with exactly the same behaviour as `File`.

An alternative option might be to keep `Tempfile` as a module and have these methods defined there, but return an ordinary `File` instance. This would reduce the breaking changes, it would just be that `Tempfile.new` returns `File` instead of `Tempfile` (etc.).

But the changes in the method signatures need to be breaking anyways in order to provide a sane API.

File: src/file/tempfile.cr (hunk 1 above)
Comment (RX14) [reply src/file/tempfile.cr]: all files use blocking IO anyway because `O_NONBLOCK` has no effect on files.

**Comment by straight-shoota** [comment]
Rebased and removed `blocking`

**Comment by straight-shoota** [comment]
@ysbaddaden Would you feel more comfortable if we put the tempfile methods back in `Tempfile` namespace (as module) and just return `File`?
This would separate tempfile methods from regular `File` methods. And it wouldn't need to be in core lib.

**Comment by RX14** [comment]
I see no reason why they can't be in the corelib, it's not particularly much more code (the code would be required as part of crystal/system anyway) and it's only dependency is libc. It doesn't drag in anything more to the corelib. This PR is fine as-is.

**Pull request merged**